metal3-scaling-experiments/create-clusters-sharded.sh

```diff
@@ -0,0 +1,171 @@
+  bmh_name="${bmh_namespace_name#*/}"
+  bmh_uid="$(kubectl -n "${namespace}" get bmh "${bmh_name}" -o jsonpath="{.metadata.uid}")"
+  sed "s/fake-node/${machine}/g" fake-node.yaml > "/tmp/${machine}-node.yaml"
+  sed -i "s/fake-uuid/${bmh_uid}/g" "/tmp/${machine}-node.yaml"
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml create -f "/tmp/${machine}-node.yaml"
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml label node "${machine}" node-role.kubernetes.io/control-plane=""
+
+    # Add static pods to make kubeadm control plane manager happy
+  sed "s/node-name/${machine}/g" kube-apiserver-pod.yaml > "/tmp/kube-apiserver-${machine}-pod.yaml"
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml create -f "/tmp/kube-apiserver-${machine}-pod.yaml"
+  sed "s/node-name/${machine}/g" kube-controller-manager-pod.yaml > "/tmp/kube-controller-manager-${machine}-pod.yaml"
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml create -f "/tmp/kube-controller-manager-${machine}-pod.yaml"
+  sed "s/node-name/${machine}/g" kube-scheduler-pod.yaml > "/tmp/kube-scheduler-${machine}-pod.yaml"
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml create -f "/tmp/kube-scheduler-${machine}-pod.yaml"
+  # Set status on the pods (it is not added when using create/apply).
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml -n kube-system patch pod "kube-apiserver-${machine}" \
+    --subresource=status --patch-file=kube-apiserver-pod-status.yaml
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml -n kube-system patch pod "kube-controller-manager-${machine}" \
+    --subresource=status --patch-file=kube-controller-manager-pod-status.yaml
+  kubectl --kubeconfig=/tmp/kubeconfig-test.yaml -n kube-system patch pod "kube-scheduler-${machine}" \
+    --subresource=status --patch-file=kube-scheduler-pod-status.yaml
+
+  # Cleanup temporary files
+  rm "/tmp/${machine}-node.yaml"
+  rm "/tmp/kube-scheduler-${machine}-pod.yaml"
+  rm "/tmp/kube-controller-manager-${machine}-pod.yaml"
+  rm "/tmp/kube-apiserver-${machine}-pod.yaml"
+}
+
+function create_shard() {
+  namespace="${1:-test-1}"
+
+  echo "Creating shard for namespace ${namespace}"
+
+  # For each shard, create a CAPKCP controller, namespace and BMHs
+  # Create namespace and BMH
+  kubectl create namespace "${namespace}"
+  kubectl -n "${namespace}" apply -f /tmp/test-hosts.yaml
+  # Create CAPKCP
+  sed "s/NAMESPACE/${namespace}/g" capkcp-deploy.yaml | kubectl apply -f -
+  # Wait for BMHs to be available
+  kubectl -n "${namespace}" wait bmh --for=jsonpath="{.status.provisioning.state}"=available --all
+}
+
+# Number of clusters per shard
+num=${1:-10}
+# Number of shards (namespaces and CAPKCP controller)
+shards=${2:-10}
+# Add more clusters in steps of step.
+step="10"
+
+# Set --namespace flag to watch specific namespace
+# Disable --leaderelect and set spec.strategy.type to Recreate
+kubectl -n capi-kubeadm-control-plane-system patch deploy capi-kubeadm-control-plane-controller-manager -p "$(cat <<EOF
```

Comment:
Can't remember the kubectl -p syntax by heart, but in case it can take input from stdin, its much cleaner to do it in `cat <<EOF .....  | kubectl -p -` style, than have multi-line sub-shell output.

Reply:
Not quite like that, but I realized it is possible with `--patch-file=/dev/stdin`.